### PR TITLE
[Fix #10522] Add new `Style/ObjectThen` cop

### LINKED ISSUE #10522: Cop idea: standardize on `#then` instead of `#yield_self`
## Is your feature request related to a problem? Please describe.

It seems like it would be nice to have a consistent style for `then` v.s. `yield_self` like there are cops for some `Enumerable` methods.

## Describe the solution you'd like

A new cop with autocorrect (maybe configurable style?) that converts `#yield_self` to `#then`. It would be unsafe to autocorrect.

## Describe alternatives you've considered

N/A

## Additional context

Not sure this is a high priority/huge need, but I thought I would share the idea. Thanks!

### PR DESCRIPTION
**Pull request by ydah**
Resolve: #10522.

This cop enforces the use of consistent method names `Object#yield_self` or `Object#then`.

## example

```ruby
# EnforcedStyle: then (default)

# bad
obj.yield_self { |x| x.do_something }

# good
obj.then { |x| x.do_something }
```

```ruby
# EnforcedStyle: yield_self

# bad
obj.then { |x| x.do_something }

# good
obj.yield_self { |x| x.do_something }
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/object_methods.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      #       then: yield_self
+      #
+      # @example
+      #   # These examples are based on the default mapping for `PreferredMethods`.
+      #
+      #   # bad
+      #   obj.yield_self
+      #
+      #   # good
+      #   obj.then
```
Comment (koic) [inline lib/rubocop/cop/style/object_methods.rb:24]: I see cases where block arguments are used.

```suggestion
      #   # bad
      #   obj.yield_self { |x| x.do_something }
      #
      #   # good
      #   obj.then { |x| x.do_something }
```

**Comment by koic** [comment]
If this design is better, `InstanceMethods` would be better named than `ObjectMethods`.

However, it may be better to implement the cop called `YieldSelf` without over-abstraction. In that case, I think user can select `EnforcedStyle: yield_self` and `EnforcedStyle: then`. Also, which one is the preferred default can be discussed in the style guide.
https://github.com/rubocop/ruby-style-guide/issues

File: lib/rubocop/cop/style/object_methods.rb (hunk 1 above)
Comment (ydah) [reply lib/rubocop/cop/style/object_methods.rb]: Oops, Sorry. I’ll add it.

File: config/default.yml [hunk 2]
```diff
@@ -5135,6 +5135,18 @@ Style/WordArray:
   # The regular expression `WordRegex` decides what is considered a word.
   WordRegex: !ruby/regexp '/\A(?:\p{Word}|\p{Word}-\p{Word}|\n|\t)+\z/'
 
+Style/YieldSelf:
+  Description: 'Preferred object methods.'
```
Comment (koic) [inline config/default.yml:5139]: https://github.com/rubocop/ruby-style-guide/pull/895 has been merged. Can you insert the style guide link?
```suggestion
  Description: 'Preferred object methods.'
  StyleGuide: '#object-yield-self-vs-object-then'
```

Comment (koic) [reply config/default.yml:5139]: And can you update text of `Description`?

Comment (ydah) [reply config/default.yml]: Thank you for your review. I've added a link to the style guide and updated the `Description`.

**Comment by ydah** [comment]
Thank you so much. I updated this PR.

**Comment by bbatsov** [comment]
I like the cop and I was about the press the "rebase" button when I realized that the name is a bit too specific in the sense that `YieldSelf` is in it. I've been wondering if there's a generic name for the abstraction that "yield_self" represents, but I can't think of anything. I can live with the current name, but if we can come up with something slightly more abstract that'd be better IMO.

**Comment by ydah** [comment]
@bbatsov Thank you so much. I couldn't think of better name than `YieldSelf` either. It may not be optimal, but given the precedent of [`StringChars`](https://docs.rubocop.org/rubocop/cops_style.html#stylestringchars), how about `ObjectYieldSelf` for naming consistency?

**Comment by ydah** [comment]
@bbatsov Following the precedent of [`Style/StringChars`](https://docs.rubocop.org/rubocop/cops_style.html#stylestringchars), I propose `Style/ObjectThen`. What do you think?

**Comment by bbatsov** [comment]
Let's go with `ObjectThen`. Thanks for working on this!

**Pull request merged**